File: py/heap_sort.py
def heapify(arr, start, end):

    root = start
    child = (2*root)+1
    while(child<=end):

        #pick the largest child
        if child+1<=end and arr[child]<arr[child+1]:
            child+=1

        #maxheap so max element should be on top
        #print "root= {} child= {}".format(arr[root], arr[child])
        if arr[root]<arr[child]:
            arr[root],arr[child] = arr[child],arr[root]
            root=child
            child = (2*root)+1
        else:
            return

def build_heap(arr, N):

    for start in range((N-2)/2,-1,-1):
        heapify(arr, start, N-1)

def heapsort(arr, N):

    for end in range(N-1, 0, -1):
        arr[end],arr[0] = arr[0],arr[end]
        heapify(arr, 0, end-1)

if __name__ == '__main__':

    arr =[2,6,3,4,1,2,9,5,8]
    #arr =[12, 3, 5, 7, 4, 19, 26]

    #creating Maxheap
    build_heap(arr, len(arr))
    heapsort(arr, len(arr))
    print arr

File: go/sorting/heap-sort.go
package main

import "fmt"

func heapify(arr []int, idx int, N int) {

	root := idx
	for root*2+1 < N {

		child := root*2 + 1

		if child+1 < N && arr[child+1] > arr[child] {
			child++
		}

		if arr[child] > arr[root] {
			arr[child], arr[root] = arr[root], arr[child]

			root = child
		} else {
			return
		}
	}
}

func sortHeap(arr []int, lidx int, ridx int) {

	for idx := ridx; idx > 0; idx-- {
		arr[0], arr[idx] = arr[idx], arr[0]
		heapify(arr[:], 0, idx)
	}

}

func buildHeap(arr []int, lidx int, ridx int) {

	//start from last iternal node
	N := ridx + 1

	for idx := (N - 2) / 2; idx >= 0; idx-- {
		heapify(arr[:], idx, N)
	}
}

func heapsort(arr []int, lidx int, ridx int) {

	buildHeap(arr[:], lidx, ridx)
	sortHeap(arr[:], lidx, ridx)
}

func main() {

	arr := [11]int{1, 4, 16, 29, 46, 53, 67, 88, 56, 43, 39}

	heapsort(arr[:], 0, len(arr)-1)

	fmt.Println(arr)

}
